File: arrays and hashing/Group Anagrams.py
from collections import defaultdict

def groupAnagrams(strs: list[str]) -> list[list[str]]:
    groups = defaultdict(list)

    for s in strs:
        chars = [0] * 26

        for c in s:
            chars[ord(c) - ord('a')] += 1

        groups[tuple(chars)].append(s)

    return list(groups.values())

File: arrays and hashing/Group Anagrams.cs
public class SolutionGroupAnagrams
{
    public IList<IList<string>> GroupAnagrams(string[] strs)
    {
        Dictionary<string, IList<string>> map = new Dictionary<string, IList<string>>();

        foreach (string s in strs)
        {
            char[] chars = s.ToCharArray();
            Array.Sort(chars);
            string sortedS = new string(chars);

            if (!map.ContainsKey(sortedS))
                map[sortedS] = new List<string>();
            map[sortedS].Add(s);
        }

        IList<IList<string>> ans = new List<IList<string>>();
        foreach (var (_, anagrams) in map)
            ans.Add(anagrams);

        return ans;
    }
}
